### PR TITLE
Specify connection when creating Allocation/Admin session for Glacier2

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR updates the IceGrid code to specify the connection when creating "plain" sessions and Admin sessions via the Glacier2 session managers hosted by IceGrid.

It's not immediately clear why we didn't do it before. I believe the connections represent the correct connections. Maybe an oversight?

See also #2756 (loosely related).

See also #2967.

### CODE REVIEW

**Comment by pepone** [comment]
> It's not immediately clear why we didn't do it before. I believe the connections represent the correct connections. Maybe an oversight?

In the original design, when a client sends a heartbeat to Glacier2, Glacier2 would send a keep-alive request to the target session. The lifetime of the session was tied to the `Client-Glacier2` connection.

In 3.7 when you register a session with a connection, the reap thread added a `HeartbeatCallbackI` which would use ACM heartbeats to keep the session alive. We didn't use a connection when registering a session created through Glacier2 because in this case the `Glacier2-IceGrid` connection is not tied to a single session.

**Comment by bernardnormier** [comment]
> The issue with the new setup is that, as long as a client keeps its session alive, the Glacier2-IceGrid connection remains open. This prevents the reaper from cleaning up any sessions created through Glacier2, as the connection doesn’t close when individual client sessions terminate.

In my view, that's not an issue.

When you create sessions using a Glacier2 router, it's the responsibility of the Glacier2 router to destroy these sessions, except in the case the Glacier2 router--IceGrid registry connection drops (in which case IceGrid cleans up sessions bound to the dropped connection).

Say we have:
client -> Glacier2 router -> IceGrid registry session manager

and the client crash. Then it's up to the Glacier2 router to "destroy" the session object hosted by IceGrid - or at least attempt to destroy it. And that's what Glacier2 router does:

https://github.com/zeroc-ice/ice/blob/f604fb280cee43034a91f920e7023dd52ba82f6b/cpp/src/Glacier2/SessionRouterI.cpp#L954

https://github.com/zeroc-ice/ice/blob/f604fb280cee43034a91f920e7023dd52ba82f6b/cpp/src/Glacier2/SessionRouterI.cpp#L740

https://github.com/zeroc-ice/ice/blob/f604fb280cee43034a91f920e7023dd52ba82f6b/cpp/src/Glacier2/RouterI.cpp#L82

File: cpp/src/IceGrid/ReapThread.cpp [hunk 1]
```diff
@@ -50,25 +50,21 @@ ReapThread::run()
             auto p = _sessions.begin();
             while (p != _sessions.end())
             {
-                try
-                {
-                    auto timestamp = p->item->timestamp(); // throws ONE if the reapable is destroyed.
+                auto timestamp = p->item->timestamp();
 
-                    if (p->timeout > 0s && (chrono::steady_clock::now() - timestamp > p->timeout))
+                if (timestamp)
```
Comment (pepone) [inline cpp/src/IceGrid/ReapThread.cpp:55]: can we do:

```
if (auto timestamp = p->item->timestamp())
{
...
```

**Review by pepone — APPROVED** [review]

**Review by externl — APPROVED** [review]

**Review by bentoi — APPROVED** [review]
The connection close callback was added when ACM heartbeats were added (in 3.6). Prior to the addition of the connection close callback, sessions were only closed based on a keep alive Slice operation being called at regular intervals.

At this time, I believe we didn't want to break compatibility and still allow 3.5 clients to work with Glacier2 3.6 or vice-versa, ditto with different IceGrid/Glacier2 versions...  

That's most likely the reason why we still have code that doesn't rely on the connection close callback for destroying sessions. And it's likely that we didn't consider removing this support with 3.7.

**Pull request merged**